planner/core/find_best_task.go

```diff
@@ -1962,30 +1962,32 @@ func (ds *DataSource) getOriginalPhysicalIndexScan(prop *property.PhysicalProper
 }
 
 func (p *LogicalCTE) findBestTask(prop *property.PhysicalProperty, planCounter *PlanCounterTp) (t task, cntPlan int64, err error) {
-	if !prop.IsEmpty() {
-		return invalidTask, 1, nil
-	}
-	if p.cte.cteTask != nil {
+	var pcte *PhysicalCTE
+	if p.cte.physicalCTE != nil {
 		// Already built it.
-		return p.cte.cteTask, 1, nil
-	}
-	sp, _, err := DoOptimize(context.TODO(), p.ctx, p.cte.optFlag, p.cte.seedPartLogicalPlan)
-	if err != nil {
-		return nil, 1, err
-	}
-
-	var rp PhysicalPlan
-	if p.cte.recursivePartLogicalPlan != nil {
-		rp, _, err = DoOptimize(context.TODO(), p.ctx, p.cte.optFlag, p.cte.recursivePartLogicalPlan)
+		pcte = p.cte.physicalCTE
+	} else {
+		sp, _, err := DoOptimize(context.TODO(), p.ctx, p.cte.optFlag, p.cte.seedPartLogicalPlan)
 		if err != nil {
 			return nil, 1, err
 		}
-	}
 
-	pcte := PhysicalCTE{SeedPlan: sp, RecurPlan: rp, CTE: p.cte, cteAsName: p.cteAsName}.Init(p.ctx, p.stats)
-	pcte.SetSchema(p.schema)
-	t = &rootTask{pcte, sp.statsInfo().RowCount, false}
-	p.cte.cteTask = t
+		var rp PhysicalPlan
+		if p.cte.recursivePartLogicalPlan != nil {
+			rp, _, err = DoOptimize(context.TODO(), p.ctx, p.cte.optFlag, p.cte.recursivePartLogicalPlan)
+			if err != nil {
+				return nil, 1, err
+			}
+		}
+
+		pcte = PhysicalCTE{SeedPlan: sp, RecurPlan: rp, CTE: p.cte, cteAsName: p.cteAsName}.Init(p.ctx, p.stats)
+		pcte.SetSchema(p.schema)
+		p.cte.physicalCTE = pcte
+	}
+	t = &rootTask{pcte, pcte.SeedPlan.statsInfo().RowCount, false}
+	if !prop.IsEmpty() {
```

Comment:
``` sql
mysql> desc with recursive cte1(c1) as (select c1 from t1 order by c1) select /*+ MERGE_JOIN(dt1, dt2) */ * from cte1 dt1 left join t1 dt2 on dt1.c1 = dt2.c1;
+--------------------------------+---------+-----------+---------------+------------------------------------------------------------+
| id                             | estRows | task      | access object | operator info                                              |
+--------------------------------+---------+-----------+---------------+------------------------------------------------------------+
| MergeJoin_9                    | 1.25    | root      |               | left outer join, left key:test.t1.c1, right key:test.t1.c1 |
| ├─Sort_21(Build)               | 4.00    | root      |               | test.t1.c1                                                 |
| │ └─TableReader_20             | 4.00    | root      |               | data:Selection_19                                          |
| │   └─Selection_19             | 4.00    | cop[tikv] |               | not(isnull(test.t1.c1))                                    |
| │     └─TableFullScan_18       | 4.00    | cop[tikv] | table:dt2     | keep order:false, stats:pseudo                             |
| └─Sort_17(Probe)               | 1.00    | root      |               | test.t1.c1                                                 |
|   └─CTEFullScan_16             | 1.00    | root      | CTE:dt1       | data:CTE_0                                                 |
| CTE_0                          | 1.00    | root      |               | Non-Recursive CTE                                          |
| └─Sort_10(Seed Part)           | 4.00    | root      |               | test.t1.c1                                                 |
|   └─TableReader_14             | 4.00    | root      |               | data:TableFullScan_13                                      |
|     └─TableFullScan_13         | 4.00    | cop[tikv] | table:t1      | keep order:false, stats:pseudo                             |
+--------------------------------+---------+-----------+---------------+------------------------------------------------------------+
11 rows in set (0.00 sec)
```

Reply:
Please add this test case.

Reply:
Done. This is not the optimial plan. Record in this [issue](https://github.com/pingcap/tidb/issues/25674)

Reply:
To fix this problem, following changes are done:
1. Change the declaration of `physicalOptimize()`, add `prop` argument.
2. Add `doOptimize()` function, add `prop` argument.
3. Add `(p *LogicalCTE) replaceSortItems` function. This function will replace columns in prop.SortItems.
4. Add `seedColMap` in `CTEClass`. Record mapping of `UniqueID` of column in CTE output schema and `original column` of seed part.